packages/backend/src/telemetry/transaction-amount.ts

```diff
@@ -0,0 +1,53 @@
+import { ValueType } from '@opentelemetry/api'
+import { ConvertError, RatesService } from '../rates/service'
+import { ConvertOptions } from '../rates/util'
+import { TelemetryService } from './service'
+import { privacy } from './privacy'
+import { Asset } from '../rates/util'
+
+export async function collectTelemetryAmount(
+  telemetryService: TelemetryService,
+  aseRates: RatesService,
+  { amount, asset }: { amount: bigint; asset: Asset }
+) {
+  if (!amount) {
+    return
+  }
+  const convertOptions = {
+    sourceAmount: amount,
+    sourceAsset: { code: asset.code, scale: asset.scale },
+    destinationAsset: {
+      code: telemetryService.getBaseAssetCode(),
+      scale: telemetryService.getBaseScale()
+    }
+  }
+  const converted = await convertAmount(
```

Comment:
My first thought seeing this was that the conversion should possibly live in the telemetry service itself so that it doesn't need to expose its own rates service outside of the telemetry class. Thoughts?

Reply:
I think that was my initial solution as well. I think i moved it out in the previous implementation on the connector side for proper unit testing (which i had trouble with there)

Giving this some more thought, we could also have another approach:
How about the existing ratesService has an optional fallback retrieval url?
since telemetry tries the ASE rates first as well,and only then tries out the external rates from coingate.

so then telemetry would not have it's own rates service.
When we need just ASE conversion, we use the rates service as usual.
When we need fallback to external rates, we just use the rates service with the flag set.

This way, telemetry itself would not have the concern of rates conversion at all, it would just use it.

Which solution would you opt for?

Reply:
@sabineschaller do you see ASEs being ok with falling back to our (external) rates in case something goes wrong on their rates API?

Reply:
> but I'm not entirely sure if an ASE would be ok with using some external rates
We would not be doing that in ASE direct use cases though. I see it as optional, and It would just be used in telemetry now, i would not enforce the fallback. It's just used when calling the `convert` method of the `ratesService` with an optional 'fallback' param

Reply:
> @sabineschaller do you see ASEs being ok with falling back to our (external) rates in case something goes wrong on their rates API?

I don't think so. That could put a big amount of risk on them. Technically, the fallback should only be used for telemetry but I don't think we can 100% guarantee that. Hence, they wouldn't want to add a fallback that could put them out of business.

Reply:
@beniaminmunteanu I understand this now:
>  I see it as optional, and It would just be used in telemetry now, i would not enforce the fallback.

@sabineschaller, why do you say 
> I don't think we can 100% guarantee that?

Reply:
I think for now let's just have two separate rates services like we have now and use them only in telemetry service (ie. do the conversion in telemetry service itself)

Reply:
> why do you say "I don't think we can 100% guarantee that"?

Because if it's a fallback, the connector could also fallback to it if the ASE's rates service is not available, couldn't it?

Reply:
> the connector could also fallback to it if the ASE's rates service is not available, couldn't it?

Yes, if we just have one rates service. I think Beni was referring to having an optional fallback URL as a parameter for the convert method, and we'd only have it set when doing the conversion the telemetry service

Reply:
Gotcha. I thought it would be a general fallback. But I'm sure I didn't read carefully before 😉

Thanks for clarifying!

Reply:
You still have a good point though about being careful. I think this is why I still prefer separate rate [services](https://github.com/interledger/rafiki/pull/2299#discussion_r1463798330)

Reply:
Separate it is then! 👌